pkg/core/consensus/reduction/firststep/reducer.go

```diff
@@ -161,11 +162,12 @@ func (r *Reducer) CollectBestScore(e consensus.Event) error {
 	lg.WithField("id", r.reductionID).Traceln("starting reduction")
 	r.startReduction()
 	step := r.eventPlayer.Forward(r.ID())
-	r.eventPlayer.Play(r.reductionID)
+	// `Play` dispatches queued events, so we should do this in a goroutine to avoid
+	// waiting too long to send a message.
+	go r.eventPlayer.Play(r.reductionID)
 
-	// sending reduction can very well be done concurrently
 	if r.handler.AmMember(r.round, step) {
-		go r.sendReduction(e.Header.BlockHash)
+		r.sendReduction(e.Header.BlockHash)
```

Comment:
While the issue is clear, there are a number of problems with your suggested solution. Apart from disliking goroutines as a tool to arbitrate instruction ordering, I see an even bigger "fixing through side effects" problem. If the issue is that queue flushing slows things down so much to impact the reduction message header, it would be better to fix that rather than introduce concurrency as a way to patch things up. In fact, I don't necessarily see a problem with the synchrony of queue flushing, as much as with these following:

```
...
// here the `round` is a state of the reducer and the `step` is local to this function and obtained by the EventPlayer/Coordinator
if r.handler.AmMember(r.round, step) {
    // in the following line, `sendReduction` calls `r.signer.Sign` which makes implicit use of the step and round the `Coordinator` is at
    // rather than the round and step of the line above
    r.sendReduction(e.Header.BlockHash) 
}
...
```

Reply:
I see your point. It's a very basic band-aid sort of fix, and the concurrency could be avoided as you point out. Do you suggest to construct the header within the `Reducer`, before passing it to the `Signer`?